### DISCUSSION TITLE
How to comment out a snippet

### DISCUSSION THREAD

**Discussion by austinliuigi** [L3MON4D3/LuaSnip#596] [Q&A] [answered]
I am creating a snippet to quickly create vim modelines, and I'm attempting to comment it out dynamically no matter what language I am expanding it in.

I am leveraging the Comment.nvim api in the snippet's callback, but ran into some issues.

```lua
local snippets = {
  s("modeline", fmt([[
    vim: set {} :
  ]], {
    i(0, "opt=value opt=value")
  }), {
    callbacks = {
      [0] = {
        [events.enter] = function()
          require("Comment.api").toggle.linewise.current()
        end
      }
    },
  }),
}
```

With this implementation, the only issue is that the text that is selected for the insert node is in its original position before the comment characters were inserted:

<img width="352" alt="image" src="https://user-images.githubusercontent.com/85013922/190034848-aa8c2dc2-ef44-48de-889e-b4623a1f800b.png">

Any tips on how to fix this?

**Comment by L3MON4D3** [comment]
Hmmm, the problem is that we [select the text](https://github.com/L3MON4D3/LuaSnip/blob/3a814e069c73d49887b708417a9f58b948e807da/lua/luasnip/nodes/insertNode.lua#L108) before [firing the event](https://github.com/L3MON4D3/LuaSnip/blob/3a814e069c73d49887b708417a9f58b948e807da/lua/luasnip/nodes/insertNode.lua#L127)...

One solution to this issue is of course to fire the event before selecting (eg. executing the "enter"), but I'm not sure we disable some other workflow with that.
Maybe we should introduce `events.enter_pre`, `events.enter_post`, and make `enter` an alias for `enter_post` (to not break current behaviour)

Until then, I think you could get away with re-using the code we use to select the insertNode (select again after toggling the comment) though I won't promise that this'll work indefinitely :P

**Reply by austinliuigi** [reply]
Sounds good, adding those extra events sounds like a good idea to give users more flexibility :)

**Comment by leiserfg** [comment, marked as answer]
You can use the `LINE_COMMENT` or the `BLOCK_COMMENT_START` `BLOCK_COMMENT_END` environment variables.

**Reply by L3MON4D3** [reply]
Wow that was pretty on-time xD
env-variables are also a nice idea :+1:
Not sure how much better `comment.nvim` is at this stuff (`comment.nvim`-env-namespace when?)

**Reply by leiserfg** [reply]
`comment.nvim` should be better only  if the comment is in a treesitter embedded language (like commenting something in a markdown code-fence), the logic behind our `*COMMENT*` was "borrowed" from https://github.com/b3nj5m1n/kommentary (I didn't credit it because I was the author of the pr that introduced that there in the first place :smile: )

**Reply by austinliuigi** [reply]
One thing about this as well is that I need to check what kind of comments a given language supports. For instance, python doesn't have block comments, whereas markdown only has block comments. Based on the current implementation of these variables, if a language doesn't support a type of comment (either line or block), then it defaults to C comment chars. Would it instead be better to default to empty strings?

**Reply by austinliuigi** [reply]
For anyone interested, this is what I put together just now and it seems to work properly:

```lua
  s("modeline", fmt([[
    {} vim: set {} :{}
  ]], {
    f(function(_, parent, _)
      local env = parent.snippet.env
      return env.LINE_COMMENT == "//" and env.BLOCK_COMMENT_START or env.LINE_COMMENT
    end),
    i(0, "opt=value opt=value"),
    f(function(_, parent, _)
      local env = parent.snippet.env
      return env.LINE_COMMENT == "//" and " " .. env.BLOCK_COMMENT_END or ""
    end),
  }))
```

**Reply by L3MON4D3** [reply]
> One thing about this as well is that I need to check what kind of comments a given language supports. For instance, python doesn't have block comments, whereas markdown only has block comments. Based on the current implementation of these variables, if a language doesn't support a type of comment (either line or block), then it defaults to C comment chars. Would it instead be better to default to empty strings?

Mhmm, not sure, but in doubt I think it's better to insert anything than nothing, that way it's at least a bit clearer where the problem originates.

**Reply by leiserfg** [reply]
It uses c style comment by default because that's what vim does (see `:h commentstring`)

**Reply by leiserfg** [reply]
We could call  `Comment.nvim` if available to get the comments from it, and fallback to our own implementation if it's not there, but I don't know if that is actually a good idea, as it will lead to different results depending on an external plugging.

**Reply by L3MON4D3** [reply]
Ah, you meant for checking... mhm maybe. Setting them to an empty string would fit nicely with lsp's var undefined or unknown. wdyt @leiserfg

**Reply by L3MON4D3** [reply]
(should've reloaded before commenting xD)
Ahh, I was confused for a second because help shows `/*%s*/` as just a bold `/%s/`. Yeah, that's a good argument, but we can easily add a namespace for `comment.nvim`s comments.

Not sure though, should it live in luasnip or comment.nvim?

**Reply by leiserfg** [reply]
> Not sure though, should it live in luasnip or comment.nvim?

A bit skeptical about integrating with 3rd parties, imagine tomorrow a better commenting plugin comes, will we force users to keep using the old one? I used to contribute/use Komentary but switched to `comment.nvim` in a blink and that could happen again.

**Reply by L3MON4D3** [reply]
Good point.. unfortunately the same argument applies to snippet engines, from a commenting-plugin-maintainer's perspective xD
The main downsides of hosting that stuff here seems to be that interface-changes can lead to broken code on our side, but there's no way around that if we want to support these integrations.

I think supporting multiple commenting-plugins shouldn't be problematic, given they all have some kind of `get_line_comment`, `get_block_comment` api.

**Reply by leiserfg** [reply]
Then we have to decide if we wanna do it as I said (check the snippet plugin the user is using and fallback to our own implementation) or have it in a different env-variable. 
As  far as I know the most used ones are `tpope/vim-commentary` (uses commentstring as we do so does not need a custom implementation)
comment.nvim  and Kommentary.
The other problem I can foresee is that some users configure their comment plugin to load on a keypress (using the keys arg of packer), that could make that the environment variable takes different values depending on the load of the plugin.

**Reply by L3MON4D3** [reply]
Oh, yeah different env-variable for sure, I was thinking of providing it via `environ_namespace`.
> The other problem I can foresee is that some users configure their comment plugin to load on a keypress (using the keys arg of packer), that could make that the environment variable takes different values depending on the load of the plugin.

Mhm yeah, that would indeed be a problem.. Not sure how to best approach this then :/
